File: scripts/EnviarAnexoCertidao/enviarAnexoCertidao.js
/**
 * Envia anexo a solicitação de certidão (EnviarAnexoCertidao) no webservice Certidões da ONR.
 */
import fs from "node:fs";
import soap from "soap";
import { resolvePath } from "../../lib/onr_env.js";
import {
  businessErrorHintAnexo,
  loadAnexoArquivo,
  resolveCertidoesProtocolo,
} from "../../lib/onr_certidoes_anexo.js";
import {
  hashErrorHint,
  loadCertidoesSoapConfig,
  loadLoginConfig,
  loadServentiaChave,
  resolveAuthHash,
} from "../../lib/onr_certidoes.js";

const PREFIX = "CERTIDOES_ENVIAR_ANEXO_";
const FIELD_ORDER = ["Hash", "Protocolo", "NomeArquivo", "ArquivoBase64"];

function loadConfig() {
  const anexo = loadAnexoArquivo(PREFIX);
  return {
    chave: loadServentiaChave(),
    loginCfg: loadLoginConfig(),
    protocolo: resolveCertidoesProtocolo(PREFIX),
    nomeArquivo: anexo.nomeArquivo,
    arquivoBase64: anexo.arquivoBase64,
    arquivoBytes: anexo.bytes,
    ...loadCertidoesSoapConfig(),
  };
}

function buildRequest(cfg, hashValue) {
  const values = {
    Hash: hashValue,
    Protocolo: cfg.protocolo,
    NomeArquivo: cfg.nomeArquivo,
    ArquivoBase64: cfg.arquivoBase64,
  };
  return Object.fromEntries(FIELD_ORDER.map((key) => [key, values[key]]));
}

function requestForLog(oRequest) {
  const b64 = oRequest.ArquivoBase64 || "";
  const preview =
    b64.length > 80 ? `${b64.slice(0, 80)}... (${b64.length} chars base64)` : b64;
  return { ...oRequest, ArquivoBase64: preview };
}

function extractResult(response) {
  return (
    response?.EnviarAnexoCertidaoResult ??
    response?.EnviarAnexoCertidaoResponse?.EnviarAnexoCertidaoResult ??
    response
  );
}

async function enviarAnexoCertidao(cfg, oRequest) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL não encontrado: ${wsdlPath}`);
  }

  const client = await soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 120000 },
  });

  let response;
  try {
    [response] = await client.EnviarAnexoCertidaoAsync({ oRequest });
  } catch (err) {
    const status = err?.response?.status;
    const body =
      typeof err?.body === "string"
        ? err.body
        : typeof err?.response?.data === "string"
          ? err.response.data
          : null;

    if (status === 503 || (body && !body.trimStart().startsWith("<"))) {
      throw new Error(
        `Servidor ONR indisponível (HTTP ${status ?? "?"}): ${body ?? err.message}`
      );
    }
    throw err;
  }

  return extractResult(response);
}

async function main() {
  const cfg = loadConfig();
  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg);
  const oRequest = buildRequest(cfg, hash);

  console.log("=== Parâmetros EnviarAnexoCertidao ===");
  console.log(JSON.stringify(requestForLog(oRequest), null, 2));
  if (cfg.arquivoBytes !== undefined) {
    console.log(`Arquivo: ${cfg.nomeArquivo} (${cfg.arquivoBytes} bytes → base64)`);
  }
  console.log(`\nEndpoint: ${cfg.endpoint}`);

  const response = await enviarAnexoCertidao(cfg, oRequest);
  console.log("\n=== Resposta ===");
  console.log(JSON.stringify(response, null, 2));

  if (!response.RETORNO) {
    console.error(
      `\nEnviarAnexoCertidao falhou: [${response.CODIGOERRO}] ${response.ERRODESCRICAO}`
    );
    const hint =
      businessErrorHintAnexo(response.CODIGOERRO) ??
      hashErrorHint(response.CODIGOERRO);
    if (hint) console.error(hint);
    process.exit(1);
  }

  console.log(
    `\nOK — Anexo ${cfg.nomeArquivo} enviado ao protocolo ${cfg.protocolo}.`
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: scripts/EnviarAnexoCertidao/enviarAnexoCertidao.py
#!/usr/bin/env python3
"""Envia anexo a solicitação de certidão (EnviarAnexoCertidao) no webservice Certidões da ONR."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from dotenv import load_dotenv

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from lib.onr_certidoes import (  # noqa: E402
    hash_error_hint,
    load_certidoes_soap_config,
    load_login_config,
    load_serventia_chave,
    resolve_auth_hash,
)
from lib.onr_certidoes_anexo import (  # noqa: E402
    business_error_hint_anexo,
    load_anexo_arquivo,
    resolve_certidoes_protocolo,
)
from lib.onr_json import to_json_safe  # noqa: E402
from lib.onr_soap import call_operation_from_cfg  # noqa: E402
from lib.onr_zeep_serialize import serialize_result  # noqa: E402

PREFIX = "CERTIDOES_ENVIAR_ANEXO_"
FIELD_ORDER = ("Hash", "Protocolo", "NomeArquivo", "ArquivoBase64")


def load_config() -> dict:
    load_dotenv(ROOT / ".env")
    anexo = load_anexo_arquivo(PREFIX)
    return {
        "chave": load_serventia_chave(),
        "login_cfg": load_login_config(),
        "protocolo": resolve_certidoes_protocolo(PREFIX),
        "nome_arquivo": anexo["nome_arquivo"],
        "arquivo_base64": anexo["arquivo_base64"],
        "arquivo_bytes": anexo.get("bytes"),
        **load_certidoes_soap_config(),
    }


def build_request(cfg: dict, hash_value: str) -> dict:
    values = {
        "Hash": hash_value,
        "Protocolo": cfg["protocolo"],
        "NomeArquivo": cfg["nome_arquivo"],
        "ArquivoBase64": cfg["arquivo_base64"],
    }
    return {key: values[key] for key in FIELD_ORDER}


def _request_for_log(o_request: dict) -> dict:
    b64 = o_request.get("ArquivoBase64") or ""
    preview = (
        f"{b64[:80]}... ({len(b64)} chars base64)"
        if len(b64) > 80
        else b64
    )
    return {**o_request, "ArquivoBase64": preview}


def main() -> int:
    cfg = load_config()
    hash_value = resolve_auth_hash(cfg["chave"], cfg["login_cfg"])
    o_request = build_request(cfg, hash_value)

    print("=== Parâmetros EnviarAnexoCertidao ===")
    print(json.dumps(_request_for_log(o_request), ensure_ascii=False, indent=2))
    if cfg.get("arquivo_bytes") is not None:
        print(
            f"Arquivo: {cfg['nome_arquivo']} ({cfg['arquivo_bytes']} bytes → base64)"
        )
    print(f"\nEndpoint: {cfg['endpoint']}")

    result = call_operation_from_cfg(cfg, "EnviarAnexoCertidao", o_request)
    response = to_json_safe(serialize_result(result))

    print("\n=== Resposta ===")
    print(json.dumps(response, ensure_ascii=False, indent=2))

    if not response.get("RETORNO"):
        codigo = response.get("CODIGOERRO", "?")
        descricao = response.get("ERRODESCRICAO", "")
        print(
            f"\nEnviarAnexoCertidao falhou: [{codigo}] {descricao}",
            file=sys.stderr,
        )
        if isinstance(codigo, int):
            hint = business_error_hint_anexo(codigo) or hash_error_hint(codigo)
            if hint:
                print(hint, file=sys.stderr)
        return 1

    print(
        f"\nOK — Anexo {cfg['nome_arquivo']} enviado ao protocolo {cfg['protocolo']}."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
